presto-main/src/main/java/io/prestosql/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -57,17 +58,19 @@
     private final CostCalculator costCalculator;
     private final List<PlanOptimizer> legacyRules;
     private final RuleIndex ruleIndex;
+    private final Predicate<Session> legacyRulesEnabled;
 
     public IterativeOptimizer(RuleStatsRecorder stats, StatsCalculator statsCalculator, CostCalculator costCalculator, Set<Rule<?>> rules)
     {
-        this(stats, statsCalculator, costCalculator, ImmutableList.of(), rules);
+        this(stats, statsCalculator, costCalculator, session -> false, ImmutableList.of(), rules);
     }
 
-    public IterativeOptimizer(RuleStatsRecorder stats, StatsCalculator statsCalculator, CostCalculator costCalculator, List<PlanOptimizer> legacyRules, Set<Rule<?>> newRules)
+    public IterativeOptimizer(RuleStatsRecorder stats, StatsCalculator statsCalculator, CostCalculator costCalculator, Predicate<Session> legacyRulesEnabled, List<PlanOptimizer> legacyRules, Set<Rule<?>> newRules)
     {
         this.stats = requireNonNull(stats, "stats is null");
         this.statsCalculator = requireNonNull(statsCalculator, "statsCalculator is null");
         this.costCalculator = requireNonNull(costCalculator, "costCalculator is null");
+        this.legacyRulesEnabled = requireNonNull(legacyRulesEnabled, "legacyRulesEnabled is null");
```

Comment:
I'll change it to `useLegacyRules`. It's not really a preference. It's a binary choice.

Reply:
Yes, from the perspective of the caller it is binary, as they provide legacy rules.

From the perspective of implementation, it is a preference, because this toggle is ignored if legacy rules is empty.

I meant the implementation perspective, but it is more important how it shows to the caller.